addons/l10n_pt/models/account_move.py

```diff
@@ -0,0 +1,278 @@
+    # MISC REQUIREMENTS
+    ####################################
+
+    def _get_starting_sequence(self):
+        # EXTENDS account sequence.mixin
+        self.ensure_one()
+        if self.company_id.account_fiscal_country_id.code != 'PT':
+            return super()._get_starting_sequence()
+        if self.journal_id.type in ['sale', 'bank', 'cash']:
+            starting_sequence = "%s%04d/00000" % (self.journal_id.code, self.date.year)
+        else:
+            starting_sequence = "%s%04d-%02d/0000" % (self.journal_id.code, self.date.year, self.date.month)
+        if self.journal_id.refund_sequence and self.move_type in ('out_refund', 'in_refund'):
+            starting_sequence = "R" + starting_sequence
+        if self.journal_id.payment_sequence and self.payment_id:
+            starting_sequence = "P" + starting_sequence
+        return starting_sequence
+
+    @api.constrains('name')
+    def _check_name(self):
+        for move in self.filtered(lambda m: (
+            m.company_id.account_fiscal_country_id.code == 'PT'
+            and m.journal_id.restrict_mode_hash_table
+            and m.name
+            and m.name != '/'
+        )):
+            if not re.match(r'^[^/^ ]+/[0-9]+$', move.name):
+                raise ValidationError(_(
+                    "The name of the document (%s) is invalid. It must start with the identifier "
+                    "of the series followed by a slash and the number of the document within the "
+                    "series (e.g. INV2024/1).", move.name
+                ))
+            prefix = move.name.split("/")[0]
+            at_series = self.env['l10n_pt.at.series'].search([('company_id', '=', move.company_id.id), ('prefix', '=', prefix)])
+            if not at_series:
+                raise ValidationError(_(
+                    "The series %s has not yet been registered with an official AT code. "
+                    "You can do so in the Accounting Settings.", prefix
+                ))
+            if at_series.type != move.move_type:
+                raise UserError(_(
+                    "The type of the series %(prefix)s (%(series_type)s) does not match the type of the document %(move_name)s (%(move_type)s).",
+                    prefix=at_series.prefix,
+                    series_type=dict(at_series._fields['type'].selection).get(at_series.type),
+                    move_name=move.name,
+                    move_type=dict(move._fields['move_type'].selection).get(move.move_type),
+                ))
+
+    @api.depends('state', 'invoice_date', 'company_id.account_fiscal_country_id.code')
+    def _compute_l10n_pt_show_future_date_warning(self):
+        for move in self:
+            move.l10n_pt_show_future_date_warning = (
+                move.company_id.account_fiscal_country_id.code == 'PT'
+                and move.state == 'draft'
+                and move.invoice_date
+                and move.invoice_date > fields.Date.today()
+            )
```

Comment:
It's recommended to add the following validation (the revisers will also ask you to test the following):

- Create a invoice now and hash it.
- Change the system date to a past period.
- Create a new invoice.
- What they expect is a block advising that there is invoices with system_entry_date (l10n_pt_hashed_on) after the current system_entry_date. 

This indicates somehow the system date was changed which is illegal for Portuguese AT.

Maybe a general validation for all hashed PT documents (stock moves for example).